berkeley-function-call-leaderboard/bfcl/model_handler/local_inference/phi_fc.py

```diff
@@ -212,17 +166,59 @@ def _extract_tool_calls(input_string: str) -> list[any]:
         Returns:
             (list) The matched objects that we were able to parse out
         """
+        # Match the normal situation: <|tool_call|>[{\"name\": \"calculate_final_velocity\", \"arguments\": {\"height\": 150, \"initial_velocity\": 0}}]<|/tool_call|>
         pattern = r"<\|tool_call\|>(.*?)<\|/tool_call\|>"
         matches = re.findall(pattern, input_string, re.DOTALL)
 
+        # Often the model will miss the `<|/tool_call|>`: <|tool_call|>[{\"name\": \"calculate_final_velocity\", \"arguments\": {\"height\": 150, \"initial_velocity\": 0}}]
+        # Since `<|tool_call|>` is still present, we consider this a valid case
```

Comment:
I do agree with this logic in general, and I believe Llama 3 also behaves this way. More generally (and not related to this PR at all) I wonder how MCP clients will handle all of this logic across different models & behaviors. Seems like a massive headache!

Reply:
Yea, this scenario is a bit tricky in my opinion. I see how people could argue both ways. In such cases, the important thing for a benchmark like BFCL is to set a consistent standard (do the right thing) and follow it. 
After careful consideration, I agree with you that, from a parsing perspective, this should be a failure scenario since the closing tags are expected but not supplied. The model is not following instructions.

Reply:
@HuanzhiMao Either way works. I can see the other perspective of -- if we can parse out the noise the model generated, is the underlying function call correct?

When I have some time this week I'll also update this PR to include results for Phi-4-mini as well (and maybe Phi-4 if I have time to get to it). I think it will be interesting to see the differences.

Reply:
That would be awesome. Thank you! I will leave this PR open for now.